### PR TITLE
Update the GULObjectSwizzler to handle NSProxy objects

### PR DESCRIPTION
**Pull request by tejasd**
A fix for: https://github.com/firebase/firebase-ios-sdk/issues/1936

GULObjectSwizzler couldn't currently handle ISA Swizzling NSProxy objects - this PR adds logic that deals with that and unit tests that would've failed if this logic weren't present.

The way this is done:
1. Stop depending on the `-[NSObjectProtocol class]` method to determine the class and use a runtime C function as the ObjC method is probably overriden to hide the fact that it is a proxy.
2. If it's a proxy, also swizzle the `-[NSObjectProtocol respondsToSelector:]` method to take into account the new methods we might've added.

Side note: It'd be so nice if there were a canonical way to access the underlying object(s) from an NSProxy.

### CODE REVIEW

File: GoogleUtilities/ISASwizzler/GULSwizzledObject.m [hunk 1]
```diff
@@ -26,6 +28,17 @@ @implementation GULSwizzledObject
 + (void)copyDonorSelectorsUsingObjectSwizzler:(GULObjectSwizzler *)objectSwizzler {
   [objectSwizzler copySelector:@selector(gul_objectSwizzler) fromClass:self isClassSelector:NO];
   [objectSwizzler copySelector:@selector(gul_class) fromClass:self isClassSelector:NO];
+
+  // This is needed because NSProxy objects ususally override
```
Comment (paulb777) [inline GoogleUtilities/ISASwizzler/GULSwizzledObject.m]: usually

File: GoogleUtilities/SwizzlerTestHelpers/GULProxy.m [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  return [_delegateObject conformsToProtocol:aProtocol];
+}
+
+- (BOOL)isProxy {
+  return YES;
+}
+
+- (NSString *)description {
+  return [_delegateObject description];
+}
```
Comment (mikehaney24) [inline GoogleUtilities/SwizzlerTestHelpers/GULProxy.m]: newline

Comment (tejasd) [reply GoogleUtilities/SwizzlerTestHelpers/GULProxy.m]: Done.

File: GoogleUtilities/ISASwizzler/GULSwizzledObject.m (hunk 1 above)
Comment (tejasd) [reply GoogleUtilities/ISASwizzler/GULSwizzledObject.m]: Done.

**Comment by paulb777** [comment]
Please update https://github.com/firebase/firebase-ios-sdk/blob/master/GoogleUtilities/CHANGELOG.md

File: GoogleUtilities/CHANGELOG.md [hunk 3]
```diff
@@ -1,6 +1,6 @@
 # Unreleased
 
-- Fixed a crash caused due to `NSURLConnection` delegates being wrapped in an
```
Comment (paulb777) [inline GoogleUtilities/CHANGELOG.md]: The original was correct https://www.englishpage.com/articles/a-vs-an.htm

**Pull request merged**